### ISSUE TITLE
[Feature]: Add callbacks to rollouts

### ISSUE THREAD

**Issue by alfieroddanintel** [open-edge-platform/physical-ai-studio#137]
### Problem Statement

The `Rollout` metric currently runs the `rollout` function using `.update`:

```python
def evaluate_gym(self, batch: Gym, batch_idx: int, stage: str) -> dict[str, float]:
    ...
    # Select the appropriate metric based on stage
    metric = self.val_rollout if stage == "val" else self.test_rollout

    # Update metric with this rollout
    metric.update(env=batch, policy=self, seed=batch_idx)
```

This means in order to add any callbacks to rollout we must do it in the `policy` or the `gym`. This is outside of the lightning philosophy.

For example to record the episodes in test I must to do something like this:

```python
import os
import imageio
from lightning.pytorch.loggers import TensorBoardLogger

from getiaction.data import LeRobotDataModule
from getiaction.eval import rollout
from getiaction.gyms import Gym, PushTGym
from getiaction.policies import ACT
from getiaction.train import Trainer


def r_callback(env: Gym, frame_buffer):
    """Collect rendered frames into list."""
    frame_buffer.append(env.render())


def save_video(frames, path, fps=10):
    imageio.mimsave(path, frames, fps=fps)
    print(f"Saved video → {path}")


if __name__ == "__main__":
    gym = PushTGym()
    datamodule = LeRobotDataModule(
        repo_id="lerobot/pusht",
        train_batch_size=128,
        val_gym=gym,
        test_gym=gym,
        num_rollouts_val=10,
        num_rollouts_test=30,
    )
    logger = TensorBoardLogger(save_dir="experiments")
    trainer = Trainer(max_epochs=300, logger=logger, check_val_every_n_epoch=10)
    policy = ACT()

    # Train & test
    trainer.fit(model=policy, datamodule=datamodule)
    trainer.test(model=policy, datamodule=datamodule)

    model_name = policy.__class__.__name__
    save_dir = os.path.join("videos", model_name)

    os.makedirs(save_dir, exist_ok=True)
    for i in range(10):
        frames = []  # new frame buffer per rollout
        rollout(
            env=gym,
            policy=policy,
            render_callback=lambda env: r_callback(env, frames)
        )
        filename = os.path.join(save_dir, f"{model_name}_rollout_{i+1}.mp4")
        save_video(frames, filename)
```

I would prefer this be a lightning callback.

### Proposed Solution

We separate out the `Rollout` metric from the actual rollout call:

```python
class PolicyModel(LightningModule):
    """
    Runs a Gym rollout and returns all data required for logging or post-processing.
    """

    def __init__(self, policy_net):
        super().__init__()
        self.policy = policy
        self.metric

    def validation_step(self, env, batch_idx):
        results = self.rollout(env)
        self.metric.update(results=results)
        return results

    def test_step(self, env, batch_idx):
        results = self.rollout(env)
        self.metric.update(results=results)
        return results
```

This has the downside of holding a large amount of memory on return from val or test step. Hopefully lightning will clean it up but we will have to test.

### Component

Library (VLA framework)

### Alternatives Considered

We could broadcast out that the rollout has started and manually inject our own callbacks:

```python
class PolicyModel(LightningModule):
    def __init__(self, policy_net):
        super().__init__()
        self.policy = policy_net

        # TorchMetric aggregator for rollout statistics
        self.val_rollout = GymRolloutMetric(sync_dist=True)     # return, length, etc.
        self.test_rollout = GymRolloutMetric(sync_dist=True)

        # Optional user-provided callbacks (e.g. logging, video, exploration stats)
        self.rollout_callbacks: list[RolloutCallback] = []

    #   Rollout execution — controlled, metric-tracked, callback-aware  
    def evaluate_gym(self, env, batch_idx, stage) -> dict[str, float]:
        """
        Run one rollout episode against `env` using current policy weights.
        Metrics are updated using TorchMetrics so aggregation works across DP/DDP.

        Callback hooks:
          on_rollout_start, on_rollout_step, on_rollout_end
        """

        metric = self._select_metric(stage)
        self._dispatch("on_rollout_start", env=env, stage=stage)

        obs, info = env.reset(seed=batch_idx)
        done, step, ep_return = False, 0, 0.0

        while not done:
            action = self.act(obs)                  # model → action
            obs, reward, terminated, truncated, info = env.step(action)

            ep_return += reward
            step += 1
            done = terminated or truncated

            self._dispatch("on_rollout_step",
                           obs=obs, action=action,
                           reward=reward, step=step,
                           done=done, stage=stage)

        metric.update(env=env, policy=self, seed=batch_idx)
        summary = metric.latest_episode()    # {sum_reward, max_reward, episode_length}

        self._dispatch("on_rollout_end", metrics=summary, stage=stage)

        prefixed = {f"{stage}/gym/{k}": float(v) for k, v in summary.items()}
        self.log_dict({f"{stage}/gym/episode/{k}": v for k, v in summary.items()},
                      on_step=True, on_epoch=False)

        return prefixed

    def validation_step(self, env, batch_idx):
        return self.evaluate_gym(env, batch_idx, stage="val")

    def test_step(self, env, batch_idx):
        return self.evaluate_gym(env, batch_idx, stage="test")

    def _select_metric(self, stage):
        return self.val_rollout if stage == "val" else self.test_rollout

    def _dispatch(self, hook, **kwargs):
        """Trigger rollout callbacks if the user attached any."""
        for cb in self.rollout_callbacks:
            fn = getattr(cb, hook, None)
            if fn is not None:
                fn(self, **kwargs)
```

I think this is weaker as it moves away from the lightning official implementations.

### Acceptance Criteria

- [ ] We can add metric and functional callbacks to our rollouts without changing policy args.
- [ ] Specifically want to implement a recording callback and a percentage sucesss metric

### Additional Context

Also closes #33
